Sources/Sentry/SentryNetworkTracker.m

```diff
@@ -101,34 +111,37 @@ - (void)observeValueForKeyPath:(NSString *)keyPath
                         change:(NSDictionary<NSKeyValueChangeKey, id> *)change
                        context:(void *)context
 {
-    if ([keyPath isEqualToString:NSStringFromSelector(@selector(state))]) {
-        NSURLSessionTask *sessionTask = object;
-        if (sessionTask.state != NSURLSessionTaskStateRunning) {
-            SentrySpanId *spanId;
-            @synchronized(sessionTask) {
-                spanId = objc_getAssociatedObject(
-                    sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID);
-                // We'll just go through once
-                objc_setAssociatedObject(sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN_ID, nil,
-                    OBJC_ASSOCIATION_RETAIN_NONATOMIC);
-            }
-
-            if (spanId != nil) {
-                id<SentrySpan> span = [self.tracker getSpan:spanId];
-                NSInteger responseStatusCode = [self urlResponseStatusCode:sessionTask.response];
-
-                if (responseStatusCode != -1) {
-                    [span setDataValue:[NSNumber numberWithInteger:responseStatusCode]
-                                forKey:@"http.status_code"];
-                }
-
-                [self.tracker finishSpan:spanId withStatus:[self statusForSessionTask:sessionTask]];
-                [sessionTask removeObserver:self forKeyPath:NSStringFromSelector(@selector(state))];
-                [SentryLog logWithMessage:@"Finished HTTP span for sessionTask"
-                                 andLevel:kSentryLevelDebug];
-            }
-        }
+    if (![keyPath isEqualToString:NSStringFromSelector(@selector(state))]) {
+        return;
+    }
+
+    NSURLSessionTask *sessionTask = object;
+    if (sessionTask.state == NSURLSessionTaskStateRunning) {
+        return;
     }
+
+    id<SentrySpan> netSpan;
+    @synchronized(sessionTask) {
+        netSpan = objc_getAssociatedObject(sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN);
+        // We'll just go through once
+        objc_setAssociatedObject(sessionTask, &SENTRY_NETWORK_REQUEST_TRACKER_SPAN, nil,
+            OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+    }
+
+    if (netSpan == nil) {
+        return;
+    }
+
+    NSInteger responseStatusCode = [self urlResponseStatusCode:sessionTask.response];
+
+    if (responseStatusCode != -1) {
+        [netSpan setDataValue:[NSNumber numberWithInteger:responseStatusCode]
+                       forKey:@"http.status_code"];
+    }
+
+    [netSpan finishWithStatus:[self statusForSessionTask:sessionTask]];
+    [sessionTask removeObserver:self forKeyPath:NSStringFromSelector(@selector(state))];
```

Comment:
`m`: I think we should also remove the observer if there is no `netSpan`, which means we should do this directly before checking if the `netSpan` is nil.

Reply:
We had this before but was giving a race condition error in some tests bruno did.
We have no way to check if sessionTask has the observer before we try to remove it.

Reply:
So we can only remove the observer if there is a `netSpan`? If so, I think we should add this as a comment.